src/backend/tests/integration/utils.py

```diff
@@ -49,3 +57,91 @@ def embed_documents(self, texts: list[str]) -> list[list[float]]:
+                f"Component of type {component_type} not found, available types: {', '.join(set(node['data']['type'] for node in self.json['data']['nodes']))}"
+            )
+        return result
+
+    def get_component_by_type(self, component_type):
+        components = self.get_components_by_type(component_type)
+        if len(components) > 1:
+            raise ValueError(f"Multiple components of type {component_type} found")
+        return components[0]
+
+    def set_value(self, component_id, key, value):
+        done = False
+        for node in self.json["data"]["nodes"]:
+            if node["id"] == component_id:
+                if key not in node["data"]["node"]["template"]:
+                    raise ValueError(f"Component {component_id} does not have input {key}")
+                node["data"]["node"]["template"][key]["value"] = value
+                node["data"]["node"]["template"][key]["load_from_db"] = False
+                done = True
+                break
+        if not done:
+            raise ValueError(f"Component {component_id} not found")
+
+
+def download_flow_from_github(name: str, version: str) -> JSONFlow:
+    response = requests.get(
+        f"https://raw.githubusercontent.com/langflow-ai/langflow/v{version}/src/backend/base/langflow/initial_setup/starter_projects/{name}.json"
+    )
+    response.raise_for_status()
+    as_json = response.json()
+    return JSONFlow(json=as_json)
+
+
+async def run_json_flow(
+    json_flow: JSONFlow, run_input: Optional[Any] = None, session_id: Optional[str] = None
+) -> dict[str, Any]:
+    graph = Graph.from_payload(json_flow.json)
+    return await run_flow(graph, run_input, session_id)
+
+
+async def run_flow(graph: Graph, run_input: Optional[Any] = None, session_id: Optional[str] = None) -> dict[str, Any]:
+    graph.prepare()
+    if run_input:
+        graph_run_inputs = [InputValueRequest(input_value=run_input, type="chat")]
+    else:
+        graph_run_inputs = []
+
+    flow_id = str(uuid.uuid4())
+
+    results, _ = await run_graph_internal(graph, flow_id, session_id=session_id, inputs=graph_run_inputs)
+    outputs = {}
+    for r in results:
+        for out in r.outputs:
+            outputs |= out.results
+    return outputs
+
+
+async def run_single_component(
+    clazz: type, inputs: dict = None, run_input: Optional[Any] = None, session_id: Optional[str] = None
+) -> dict[str, Any]:
+    component = clazz(**(inputs or {}))
+
+    flow_id = str(uuid.uuid4())
+    graph = Graph(user_id=str(uuid.uuid4()), flow_id=flow_id)
+    component_id = graph.add_component(component)
+    graph.prepare()
+    if run_input:
+        graph_run_inputs = [InputValueRequest(input_value=run_input, type="chat")]
+    else:
+        graph_run_inputs = []
+
+    _, _ = await run_graph_internal(
+        graph, flow_id, session_id=session_id, inputs=graph_run_inputs, outputs=[component_id]
+    )
+    return graph.get_vertex(component_id)._built_object
```

Comment:
The dunder call on the Component class runs the Component using the inputs passed through a `set` call. Maybe we should build the tests around that. What do you think?

Reply:
I saw that but I don't quite understand the reason. We should mimic the graph run as much as possible. For example, depending on the flow structure I've noticed that some parameters are not passed correctly (see the fix of double vertices sort in this PR). I think it's better to test the code most similar to the one used by the API endpoints

Reply:
Agree